RavenDB-23453 Vector search filter - Corax's part.

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23453

### Additional description

_Note: this PR contains only the low-level implementation of filter capabilities inside Corax and Hnsw. It does **not** integrate them into RavenDB querying yet._

This PR allows passing a filter query as a condition that the results from `VectorSearch` must satisfy.

It includes:

- **Scanning query (for certain filter sizes):** retrieves node IDs from documents and performs `ExactSearch` only on those specific nodes. This is important because we must ensure that the `NumberOfCandidates` parameter is honored.  
- **Exact:** creates a mapping of all nodes and then, since it is sorted, yields nodes from best to worst.  
- **NearestSearch:** over-fetches (by increasing `NumberOfCandidates` internally). When we need more vectors (because previous ones are rejected by the filter), we increase `NumberOfCandidates` internally and restart the whole query. Thanks to this, we don’t need to implement an evicted list and handle it manually. To make this more performant, we leverage `SearchState` and its caching capabilities and store `D(VectorQuery, Node)` inside the nodes, so when we need to visit it again, we don’t read the vector from disk or recompute the distance.

Additionally, it includes refactoring of the flow:

- `VectorSearch` is now lazily initialized, meaning it will not search until we actually call `Fill` (or `AndWith`).  
- The implementation is split into multiple files to improve readability.  
- Internal searchers are now implemented as `IEnumerable`s, which means we can easily “fetch more” results.


### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [x] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-23453 Moves GrowableBitArray into Sparrow project.

maciejaszyk committed:
RavenDB-23453 Adjusted tests and API of classes required for VectorSearch filter.

maciejaszyk committed:
RavenDB-23453 Filtered Vector Search: implement filtering inside VectorSearch.

Modifications:
- Scanning mode: translate document IDs from the filter into nodes and perform ExactSearch on them.
- ExactSearch: when a filter is provided, keep all node statistics so we can return results from best to worst that fulfill the filter conditions.
- NearestSearch: leverage the stateful (cached) SearchState and rerun the query with a larger NumberOfCandidates to hopefully find nodes that are accepted by the filter conditions.

redknightlois commented on test/SlowTests/Corax/RavenDB-23453.cs:
```diff
@@ -0,0 +1,422 @@
+﻿using System;
```
Naming for the test seems a bit off based on our usual standards. Also issue ID is missing a number.

maciejaszyk replied:
Renamed.

redknightlois commented on test/SlowTests/Corax/RavenDB-23453.cs:
```diff
@@ -0,0 +1,422 @@
+    [InlineDataWithRandomSeed(true, true)]
+    public void CanFilterVectors(bool shouldScan, bool isExact, int seed)
+    {
+        var (resultQueryContainsAllDb, resultsFromFilteredQuery, resultsWithoutFilterQuery, manualQuery, docs) = CanFilterVectorsBase(shouldScan, isExact, seed);
```
Would be great to have a comment showing what type of operation we are executing against here. Thinking in terms of what would happen if yourself look at this code 6 months from now. CanFilterVectorsBase has intrincate logic that is not easily followable. The same applies to the other tests.

maciejaszyk replied:
Added explanation in `CanFilterVectorsBase`.

redknightlois commented on test/SlowTests/Corax/GrowableBitArrayTests.cs:
```diff
@@ -6,6 +6,7 @@
 using FastTests;
 using FastTests.Voron.FixedSize;
 using Sparrow.Server;
+using Sparrow.Server.Collections;
```
Do we need this import?

maciejaszyk replied:
Yes, I've moved GrowableBitArray into Sparrow.Server. Since Hnsw is inside Voron, we need to have access to it in the Corax and Voron projects as well.

redknightlois commented on src/Voron/Util/ContextBoundNativeList.cs:
```diff
@@ -19,6 +20,12 @@ public ContextBoundNativeList(ByteStringContext ctx, int requestedSize = 0)
                 Inner.Initialize(ctx, requestedSize);
         }
 
+        public IEnumerable<T> Iterate()
```
Use an struct enumerable here. The performance implication of using `IEnumerable<T>` even under devirtualization constraints is something we should take care of. The point is to avoid using the interface (and therefore boxing) only when we need to.

maciejaszyk replied:
Done

redknightlois commented:
I am holding off on reviewing for the time being. I have seen many potentially performance troubling constructs. Let's ensure first that the implementation does not regress so hard that it needs to be redesigned first.

maciejaszyk committed:
RavendB-23453 GrowableBitArray: changed iterator into struct to avoid unnecessary allocations.

maciejaszyk committed:
RavenDB-23453
- Assign `QueryPlan` only after at least one `Fill` call.
- Reuse already calculated distances in `VectorSearchRetriever`.

maciejaszyk committed:
RavenDB-23453 Added benchmark for vector search (without filter).

maciejaszyk commented on bench/Voron.Benchmark/Corax/VectorSearchBenchmark.cs:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    [Benchmark]
+    public int VectorQuery()
```
cc @redknightlois 

v7.1 stable:

| Method      | Size  | Mean     | Error    | StdDev   |
|------------ |------ |---------:|---------:|---------:|
| VectorQuery | 360   | 11.67 us | 0.230 us | 0.307 us |
| VectorQuery | 30000 | 33.79 us | 0.284 us | 0.266 us |



This PR:

| Method      | Size  | Mean     | Error    | StdDev   |
|------------ |------ |---------:|---------:|---------:|
| VectorQuery | 360   | 12.29 us | 0.107 us | 0.090 us |
| VectorQuery | 30000 | 29.57 us | 0.264 us | 0.247 us |

Lwiel commented on src/Corax/Querying/Matches/MultiVectorSearchMatch.cs:
```diff
@@ -4,128 +4,201 @@
+
+    private GrowableBitArray _filterResults;
+    private IQueryMatch _filterQuery;
+    private bool _filterQueryLoaded;
```
Check warnings

maciejaszyk replied:
done

Lwiel commented on src/Corax/Querying/Matches/MultiVectorSearchMatch.cs:
```diff
@@ -134,26 +207,41 @@ public void Score(Span<long> matches, Span<float> scores, float _)
 
                 var distance = _distances.Results[pos];
-                scores[i] += _nearestSearches[0].DistanceToScore(distance);
+                scores[i] += boostFactor * _vectorsRetrievers[0].DistanceToScore(distance);
```
Why we only do this for the first retriever?

maciejaszyk replied:
All `VectorRetrievers` were built with exactly the same scoring method, so the method is identical for all of them. I just used the first one.

arekpalinski replied:
Previously we didn't have `boostFactor` at all. Was it a bug?

Lwiel commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -35,57 +52,115 @@ public struct VectorSearchMatch : IQueryMatch
+    /// <summary>
+    /// Initialization of vector search is lazy to avoid expensive computation/IO during QueryBuilding phase.
+    /// </summary>
+    /// <param name="tmpBuffer">Used to fetch data from a filter match. This way we can spare on allocations, since the actual data will be stored in bitmap.</param>
```
We don't have this parameter

maciejaszyk replied:
Leftover, removed.

Lwiel commented on src/Voron/Data/Graphs/Hnsw.NearestSearcher.cs:
```diff
@@ -0,0 +1,209 @@
+                        // because we keep distances in memory from the previous run.
+                        // This method can be greedy enough to traverse the entire graph, so it's the caller's
+                        // responsibility to enforce a stop condition.
+                        goto Start;
```
Shouldn't we bump `_internalNumberOfCandidates` in such case?

maciejaszyk replied:
Increasing the number of candidates is the caller's responsibility, not the searcher's. Because the searcher uses a greedy algorithm without broader context (e.g., filter population), the stop condition and NoC incrementation logic were moved to the outer scope.

See:
`IncreaseNumberOfCandidates`

maciejaszyk committed:
RavenDB-23453 Increase NumberOfCandidates for NearestSearch due to filter internally.

Lwiel commented on src/Voron/Data/Graphs/Hnsw.Node.cs:
```diff
@@ -0,0 +1,117 @@
+            if ((VectorId & Constants.Graphs.VectorStorage.VectorContainerInternalIndexer) == 0)
+                return VectorId;
+
+            return VectorId & ~0xFFF;
```
Do we have a Voron const for this value?

arekpalinski replied:
ping

Lwiel commented on src/Voron/Data/Graphs/Hnsw.VectorSearchRetriever.cs:
```diff
@@ -0,0 +1,298 @@
+                        break;
+                    }
+
+                    Console.WriteLine($"CP: {_vectorsSearcher.CandidatesProcessed}");
```
Just a note to remove this

redknightlois commented on bench/Voron.Benchmark/Corax/VectorSearchBenchmark.cs:
```diff
@@ -0,0 +1,85 @@
+        var y = MathF.Sin(180 * 2 * MathF.PI / 360);
+        float[] vec = [x, y];
+        var v = GenerateEmbeddings.FromArray(indexSearcher.Allocator, vec, Raven.Client.Documents.Indexes.Vector.VectorEmbeddingType.Single, Raven.Client.Documents.Indexes.Vector.VectorEmbeddingType.Single);
+        var vs = indexSearcher.VectorSearch(_mapping.GetByFieldId(2).Metadata, v, 0.0f, 16, false, false);
```
The benchmark numbers are dominated by preparation work, you need to move as much as you can away from the benchmark itself.

redknightlois commented on bench/Voron.Benchmark/Corax/VectorSearchBenchmark.cs:
```diff
@@ -0,0 +1,85 @@
+            {
+                var id = $"doc/{i}";
+                _ids.Add(id);
+                using var builder = indexWriter.Index(id);
```
I would change this code to create clusters using high dimensional gaussian splats.  The rationale is that by moving the dimensions along `i` you are creating a situation where the data is in a diagonal; which is usually a non common scenario. Creating clusters along a number of gaussians in high dimensions is closer to reality.

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -6,27 +6,44 @@
 public struct VectorSearchMatch : IQueryMatch
 {
-    private Hnsw.NearestSearch _nearestSearch;
+    private const int ScanningThreshold = 1024;
```
should we make this configurable at the index level?

maciejaszyk replied:
Yes. This is going to be configurable in the PR related to RavenDB changes.

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -35,57 +52,115 @@ public struct VectorSearchMatch : IQueryMatch
+        if (_isEmpty)
+            return 0;
+        
+        if (_resultsPersisted == false)
+            FillAndPersistResults();
+
```
Here, shouldn't we also do a scan, basically?
What does `AndWith` _means_ here? 


`where Category = 'Fun' and vector.search(Name, $q)`

That basically say: sort the results based filtered by category, no ?

maciejaszyk replied:
The idea behind the filter is based on the QueryBuilder. In #21737, we are reordering AND conditions that are on the same level, like:
```
where [...] and VectorSearch() and [...]
```
This essentially rewrites the query into `VectorSearch(..., condition0 and condition and [...])`


However, for queries like this:
```
where Condition1 and boost(VectorSearch() and Condition_2, 3f)
```

we cannot put `Condition1` into the vector search clause with the current scoring mechanism. In such cases, we will likely perform `boost(VectorSearch() and Condition_2, 3f).AndWith(results_from_condition1)`.
I think this is acceptable for such cases; however, we must document exactly how the vector search filter works.

ayende commented on src/Voron/Data/Graphs/Hnsw.ExactSearcher.cs:
```diff
@@ -0,0 +1,112 @@
+            {
+                var pq = new PriorityQueue<long, float>();
+                Span<byte> vector = _vector.Span;
+                IEnumerable<long> toScan = _nodesToScan.HasValue ? _nodesToScan.Value.Iterate() : AllNodes();
```
elegant

ayende commented on src/Voron/Data/Graphs/Hnsw.ExactSearcher.cs:
```diff
@@ -0,0 +1,112 @@
+                _vector = vector;
+                _hasFilterMatch = hasFilterMatch;
+                //TODO: more than int32 vectors
+                PortableExceptions.ThrowIf<NotSupportedException>(_searchState.Options.CountOfVectors >= int.MaxValue && hasFilterMatch, "Cannot have more than int32 vectors and filter match");
```
```suggestion
                PortableExceptions.ThrowIf<NotSupportedException>(_searchState.Options.CountOfVectors >= int.MaxValue && hasFilterMatch, $"Cannot have more than {int.MaxValue} vectors and filter match");
```

ayende commented on src/Voron/Data/Graphs/Hnsw.NearestSearcher.cs:
```diff
@@ -0,0 +1,209 @@
+                _vector = vector;
+                _level = level;
+                NumberOfCandidates = numberOfCandidates;
+                _internalNumberOfCandidates = hasFilterMatch ? (int)Math.Min((long)int.MaxValue, (2 * (long)numberOfCandidates)) : numberOfCandidates;
```
what if we have 1M candidates in the filter match? That would be very expensive, no?

maciejaszyk committed:
RavenDB-23453
- Dispose of allocations made by SearchState
- Benchmark based on actual embeddings
- NativeList: Reset Count to 0 upon disposal
- PR comments

ayende commented on src/Voron/Data/Graphs/Hnsw.NearestSearcher.cs:
```diff
@@ -0,0 +1,209 @@
+                _vector = vector;
+                _level = level;
+                NumberOfCandidates = numberOfCandidates;
+                _internalNumberOfCandidates = hasFilterMatch ? (int)Math.Min((long)int.MaxValue, (2 * (long)numberOfCandidates)) : numberOfCandidates;
```
I don't follow how filtering works here?

maciejaszyk replied:
Discussed offline.
Actual filtering happens in `VectorSearchRetriever`

arekpalinski commented on src/Corax/Querying/Matches/MultiVectorSearchMatch.cs:
```diff
@@ -4,128 +4,196 @@
+        _scanningQuery = IndexSearcher.VectorSearchUtils.ShouldScan(_indexSearcher, _filterMatchesCount, _isExact, _filterQuery, _scanningThreshold);
+        ContextBoundNativeList<long> nodesIdsToScan = default;
+        if (_scanningQuery)
+            IndexSearcher.VectorSearchUtils.TryConvertDocumentsIdsToNodesIds(_indexSearcher, _metadata, _filterResults, out nodesIdsToScan);
```
What if `TryConvert...` will return false?

maciejaszyk committed:
RavenDB-23453
- VectorSearchMatch is empty when we have not found any nodes to scan.
- Addressed PR comments.

maciejaszyk committed:
RavenDB-23453 Rebase fixes

maciejaszyk committed:
RavenDB-23453
- Added `ShouldContinueSearch` to manage search termination based on filter document count.
- Improved candidate prefetch size calculation for `NearestSearcher`.

maciejaszyk committed:
RavenDB-23453
- Stop condition in post-filter VectorSearch is based on I/O
- Fix QueryPlan when filter is empty (vector search is not be initialized in such scenario).

maciejaszyk committed:
RavenDB-23453 Initial nodes in approximate search come from a filter.

maciejaszyk committed:
RavenDB-23453
- Introduced `Random` support for filtered vector search to enable deterministic testing.
- Adjusted `ShouldContinueSearch` logic for small filter sets.

maciejaszyk committed:
RavenDB-23453 Exact search: Cache distance calculations and optimize I/O operations.

arekpalinski commented on src/Corax/Querying/IndexSearcher.VectorSearch.cs:
```diff
@@ -7,64 +7,162 @@
+                    {
+                        if (nodesByVectorId.TryGetValue(vectorId, out var nodeId))
+                            yield return nodeId;
+                    }
```
What is the meaning of `vectorsByHash.TryGetValue()` returning `false`?

maciejaszyk replied:
That would mean the data is somehow corrupted. I used TryGetValue since it's the only exposed method. I will add a DebugAssert for that.

arekpalinski commented on src/Sparrow.Server/Collections/GrowableBitArray.cs:
```diff
@@ -0,0 +1,296 @@
+        }
+    }
+
+    public static IEnumerable<long> Probe(GrowableBitArray filterResults, Random shared)
```
Why `shared` name?

maciejaszyk replied:
Leftover, changed.

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -290,6 +121,8 @@ private static Slice SliceFromString(LowLevelTransaction llt, string name)
             return slice;
         }
+        
+        public Lookup<Int64LookupKey> NodeIdsByVectorId => _tree.LookupFor<Int64LookupKey>(Hnsw.NodesByVectorIdSlice);
```
The usage of `=>` will cause that that `LookupFor()` will be evaluated every single time you access `NodeIdsByVectorId`. Is that indented?

maciejaszyk replied:
Yes, I cached the reference locally.

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -484,8 +318,29 @@ public float Distance(ReadOnlySpan<byte> vector, int fromIdx, int toIdx)
+            return distance;
+        }
+
+        //Allows storing cached distance to the queried vector. Should be used only in querying part!
```
> Should be used only in querying part!

Are we able to have `Debug.Assert` to ensure that?

maciejaszyk replied:
We're asserting it inside `Fill` methods (class: `VectorSearchRetriever`) to ensure we're reusing the values.

arekpalinski replied:
Great

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -515,105 +370,74 @@ public enum NearestEdgesFlags
+            bool hasFilterMatch)
+        {
+            // in case of filtered match we want to over-fetch. 
+            return new NearestSearcher(this, startingPointIndex, vector, level , numberOfCandidates, candidates, flags, hasFilterMatch);
```
What part of the code indicates that we're gonna "over-fetch" as the above comment says?

maciejaszyk replied:
https://github.com/maciejaszyk/ravendb/blob/3abd101d0925152f4739697c2e185781ccad4ba3/src/Voron/Data/Graphs/Hnsw.NearestSearcher.cs#L47-L49

arekpalinski replied:
Can you somehow reference to that in the comment so it will be easier for someone to follow the flow

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -515,105 +370,74 @@ public enum NearestEdgesFlags
-                candidate.Visited = visitedCounter;
-             
-                ref var edges = ref candidate.EdgesPerLevel[level];
+                if (nodes.Count < maxBest)
```
What is `maxBest`? It appeared to me a bit unusual that we compare `Count` with some max value. Can you please explain that?

maciejaszyk replied:
MaxBest is how many nodes we want to retrieve from our filter set. I've changed the naming.

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -515,105 +370,74 @@ public enum NearestEdgesFlags
-                ref var edges = ref candidate.EdgesPerLevel[level];
+                if (nodes.Count < maxBest)
+                {
+                    // MinPriorityQueue. First to remove is a node from lower layers
```
Where do you actually do the removal? I don't see that

maciejaszyk replied:
EnqueueDequeue handles that internally.

arekpalinski replied:
I see

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -515,105 +370,74 @@ public enum NearestEdgesFlags
-                    
-                    Debug.Assert(_candidatesQ.Count > 0);
-                    _nearestEdgesQ.TryPeek(out _, out lowerBound);
+                    nodes.EnqueueDequeue(nodeId, -maxNodeLevel);
```
1. Shouldn't the above comment `// MinPriorityQueue. First to remove is a node from lower layers` be here, as this is where you remove the minimal element?

2. Regarding the actual removal, the minimal element that will be removed is actually the maximum node level. Can you please double check that you're doing what you wanted?

3. I wonder if we shouldn't provide a custom comparer to `PriorityQueue` instead of dealing with negative values hence it's more difficult to follow the code.

maciejaszyk replied:
I rewrote this code and added more comments.

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.NearestSearcher.cs:
```diff
@@ -0,0 +1,303 @@
+                var max = filterDocsCount switch
+                {
+                    < 1024 => 2 * filterDocsCount, // Todo: it's hard to determine right number here. For smaller graphs it may be an issue.
+                                                        // However, for small filter set (for 1K we will force exact search anyway, therefore this probably will be used only for testing purposes.
```
What about this TODO? I see it's like that in the PR integrating this to RavenDB as well.

ayende commented on src/Corax/Querying/Matches/MultiVectorSearchMatch.cs:
```diff
@@ -1,137 +1,224 @@
 
-    public QueryCountConfidence Confidence => QueryCountConfidence.Low;
-    public bool IsBoosting { get; }
+        _isEmpty = allEmpty || (_filterQuery != null && _filterResults.Count == 0);
```
why do we have to do _any_ work if `_filterResults.Count == 0` ? 
can we just return `[]` and be done with it?

maciejaszyk replied:
Yes, fixed

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -35,88 +54,168 @@ public struct VectorSearchMatch : IQueryMatch
+        };
+        
 
+        _isEmpty = _scanningQuery 
```
same here, if `_filterMatchesCount == 0 ` - we can avoid a lot of work

maciejaszyk replied:
Fixed as well

ayende commented on src/Sparrow.Server/Collections/GrowableBitArray.cs:
```diff
@@ -0,0 +1,296 @@
+        }
+    }
+
+    public static IEnumerable<long> Probe(GrowableBitArray filterResults, Random shared)
```
```suggestion
    public static IEnumerable<long> Probe(GrowableBitArray filterResults, Random random)
```

ayende commented on src/Sparrow.Server/Collections/GrowableBitArray.cs:
```diff
@@ -0,0 +1,296 @@
+            if (seen.Add(current) == false)
+                continue;
+            if (filterResults.Contains(current) == false)
+                continue;
```
You are generating N random numbers.
Then you _may_ have them in your filter, but maybe not?

Wouldn't that _usually_ lead to no entries?

maciejaszyk replied:
I rewrote the enumerator to be more deterministic. Please review again.

maciejaszyk committed:
RavenDB-23453 Refactor HNSW search logic

maciejaszyk committed:
RavenDB-23453  Implement RandomNodesFromFilterEnumerator for efficient random node selection from filter results
